enhancements/installer/storage-class-encrypted.md

```diff
@@ -0,0 +1,306 @@
+---
+title: storage-class-encrypted
+authors:
+  - "@patrickdillon"
+reviewers:
+  - "@jhixson74, installer"
+  - "@jsafrane, storage"
+  - "@bertinatto, storage"
+  - "@trilokgeer, storage"
+approvers:
+  - "@jsafrane"
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - "@joelspeed"
+creation-date: 2022-06-22
+last-updated: 2022-12-09
+tracking-link:
+  - https://issues.redhat.com/browse/CORS-2049
+  - https://issues.redhat.com/browse/OCPPLAN-9307
+---
+
+# Default Storage Class with Encrypted Keys 
+
+## Summary
+
+When user-managed encryption keys are provided at install time, the default storage
+class should be set to utilize those keys. Presently, only root volumes for nodes are
+encrypted with those keys--but users would expect the persistent storage volumes to
+be encrypted as well. This enhancement proposes a solution for AWS, GCP, & Azure.
+
+## Motivation
+
+The primary motivation is to enable managed services to encrypt storage volumes with
+customer-provided keys automatically as part of the installation process without requiring
+day-2 operations. The behavior of automatically setting the default storage class is
+sensible for individual (non-managed-service) users as well; setting the default will
+reduce their day-2 burden and decrease the opportunity for user error.
+
+### User Stories
+
+When I install a cluster, I want to provide user-managed encryption keys in the install-config
+so that the default storage class will be set to provision volumes with those keys.
+
+### Goals
+
+* Default storage class is set to use user-provided keys at cluster creation for public clouds (AWS, GCP, Azure)
+
+### Non-Goals
+* Other clouds besides AWS, GCP, & Azure are out of scope
+
+
+## Proposal
+
+### Workflow Description
+
+1. Cluster creator specifies (aws|gcp|azure) user-managed encryption keys 
+in `defaultMachinePlatform` of install config.
+2. Installer creates a manifest for `ClusterCSIDriver` which includes encryption keys in spec.
+3. Cluster-storage-operator/CSI drivers use installer-created `ClusterCSIDriver`
+to create default storage class using user-encrypted keys.
+
+
+### API Extensions
+
+This enhancement extends the [`ClusterCSIDriver` API object](https://github.com/openshift/api/blob/efeef9d83325a0ba12a636e5efd7d9414036f9f7/operator/v1/types_csi_cluster_driver.go).
+
+AWS, Azure & GCP will be added as new valid values for the `CSIDriverType` union discriminator
+in the CSIDriverConfigSpec:
+
+```go
+// CSIDriverType indicates type of CSI driver being configured.
+// +kubebuilder:validation:Enum="";vSphere
+type CSIDriverType string
+
+const (
+  AWSDriverType     CSIDriverType = "aws"
+  AzureDriverType   CSIDriverType = "azure"
+  GCPDriverType     CSIDriverType = "gcp"
+  VSphereDriverType CSIDriverType = "vSphere" // existing platform
+)
```

Comment:
This makes sense. 

Perhaps the CSO will need to change some existing code with the `VSphereDriverType` -> `LegacyVSphereDriverType` change so cc @jsafrane for awareness on this point.

Reply:
The vSphere change could be a separate thing done later if need be, but the end goal should be a consistent set as I've described in the comment above

Reply:
> Why is the empty string included if the field is required?

what about upgraded clusters? They won't have this value set. I think that is why - empty string was kept. 

> VSphereDriverType CSIDriverType = "VSphere"
  LegacyVSphereDriverTpye CSIDriverType = "vSphere"

We already GAed this and since this is user editable - we will have to always support this value. Does it make sense to carry both values forever?

Reply:
> what about upgraded clusters? They won't have this value set. I think that is why - empty string was kept.

Where is this in openshift/api? 

> We already GAed this and since this is user editable - we will have to always support this value. Does it make sense to carry both values forever?

For consistency, I would argue yes, but I'm not going to force you to do it. The new values should definitely be correctly cased though

Reply:
https://github.com/openshift/api/blob/master/operator/v1/types_csi_cluster_driver.go#L109